'meson test' fails to run on Python 3.12

**Describe the bug**
Meson fails to run on Python 3.12 due to reliance on the implicit instantiation of the asyncio mainloop that was deprecated in Python 3.10. It has now been removed.

Additional details in [this Fedora bug](https://bugzilla.redhat.com/show_bug.cgi?id=2155490) and [this Python upstream issue](https://github.com/python/cpython/issues/93453).


**To Reproduce**

* Install python 3.12 alpha.
* Run `meson test` in any project that supports it


**Expected behavior**
Tests should run

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
  * Native
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
  * Experienced on Fedora Rawhide (38), but likely present on any Linux with Python 3.12
* what Python version are you using e.g. 3.8.0
  * Python 3.12
* what `meson --version`
  * Meson 0.64.1
* what `ninja --version` if it's a Ninja build
  * Ninja 1.5.8.1

I reran with Meson master branch and Python 3.12.0a3 I built myself on CentOS.

```sh
$ /local/bin/python3 run_unittests.py

Meson build system 1.0.0.rc2 Unit Tests
=========================================== test session starts
===========================================platform linux -- Python 3.12.0a3, pytest-7.2.0, pluggy-1.0.0
rootdir: /code/meson, configfile: setup.cfg
collected 479 items
```

```
========================================= short test summary info 
FAILED run_unittests.py::AllPlatformTests::test_scripts_loaded_modules - AssertionError: Lists differ: ['mesonbuild', 'mesonbuild._pathlib', 'meso[173 chars]ore'] != ['__edita...
FAILED run_unittests.py::AllPlatformTests::test_prebuilt_object - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_long_output - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_prebuilt_shared_lib_rpath - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_testsetup_default - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_force_fallback_for_nofallback - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_lto_threads - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_force_fallback_for - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_reconfigure - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_dist_git_script - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_static_library_lto - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_prebuilt_shared_lib - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_testsetup_selection - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_prebuilt_static_lib - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_forcefallback - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_suite_selection - AssertionError: 2 != 4
FAILED run_unittests.py::AllPlatformTests::test_dist_git - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_testsetups - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_testrepeat - subprocess.CalledProcessError: Command '['/code/cmake-python-build/build/local/bin/python3',...
FAILED run_unittests.py::CrossFileTests::test_cross_exe_passed_no_wrapper - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_pkgconfig_static - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::AllPlatformTests::test_wrap_git - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::CrossFileTests::test_needs_exe_wrapper_true_wrapper - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::LinuxlikeTests::test_freezing - AssertionError: 'Traceback' unexpectedly found in 'ninja explain: output meson-benchmark-prereq of pho...
FAILED run_unittests.py::CrossFileTests::test_needs_exe_wrapper_false - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::LinuxlikeTests::test_symlink_builddir - subprocess.CalledProcessError: Command '['/code/cmake-python-build/build/local/bin/python3',...
FAILED run_unittests.py::SubprojectsCommandTests::test_purge - subprocess.CalledProcessError: Command '['/code/cmake-python-build/build/local/bin/python3',...
FAILED run_unittests.py::CrossFileTests::test_needs_exe_wrapper_true - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::SubprojectsCommandTests::test_foreach - subprocess.CalledProcessError: Command '['/code/cmake-python-build/build/local/bin/python3',...
FAILED run_unittests.py::LinuxlikeTests::test_usage_external_library - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::LinuxlikeTests::test_pch_with_address_sanitizer - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::LinuxlikeTests::test_link_arg_fullname - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::LinuxlikeTests::test_static_link - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::LinuxlikeTests::test_pkgconfig_uninstalled - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
FAILED run_unittests.py::SubprojectsCommandTests::test_git_update - subprocess.CalledProcessError: Command '['/code/cmake-python-build/build/local/bin/python3',...
FAILED run_unittests.py::AllPlatformTests::test_verbose - subprocess.CalledProcessError: Command '['/code/cmake-python-build/build/local/bin/python3',...
FAILED run_unittests.py::LinuxlikeTests::test_wrap_with_file_url - subprocess.CalledProcessError: Command '['/ninja-1.11.1/ninja', '-w', 'dupbuild=err', '-d', ...
=================== 37 failed, 370 passed, 72 skipped, 451 warnings in 77.25s (0:01:17) ===================Total time: 77.540 seconds
```

--- a/mesonbuild/mesonmain.py
+++ b/mesonbuild/mesonmain.py
@@ -235,6 +235,13 @@ def set_meson_command(mainfile):
     mesonlib.set_meson_command(mainfile)
 
 def run(original_args, mainfile):
+    if os.environ.get('MESON_SHOW_DEPRECATIONS'):
+        # workaround for https://bugs.python.org/issue34624
+        import warnings
+        for typ in [DeprecationWarning, SyntaxWarning, FutureWarning, PendingDeprecationWarning]:
+            warnings.filterwarnings('error', category=typ, module='mesonbuild')
+        warnings.filterwarnings('ignore', message=".*importlib-resources.*")
+
     if sys.version_info >= (3, 10) and os.environ.get('MESON_RUNNING_IN_PROJECT_TESTS'):
         # workaround for https://bugs.python.org/issue34624
         import warnings

--- a/mesonbuild/msubprojects.py
+++ b/mesonbuild/msubprojects.py
@@ -699,7 +699,8 @@ def run(options: 'Arguments') -> int:
             raise MesonException(f'Unknown subproject type {t!r}, supported types are: {ALL_TYPES_STRING}')
     tasks: T.List[T.Awaitable[bool]] = []
     task_names: T.List[str] = []
-    loop = asyncio.get_event_loop()
+    loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(loop)
     executor = ThreadPoolExecutor(options.num_processes)
     if types:
         wraps = [wrap for wrap in wraps if wrap.type in types]

--- a/mesonbuild/mtest.py
+++ b/mesonbuild/mtest.py
@@ -1925,9 +1925,12 @@ class TestHarness:
     def run_tests(self, runners: T.List[SingleTestRunner]) -> None:
         try:
             self.open_logfiles()
-            # Replace with asyncio.run once we can require Python 3.7
-            loop = asyncio.get_event_loop()
-            loop.run_until_complete(self._run_tests(runners))
+
+            # TODO: this is the default for python 3.8
+            if sys.platform == 'win32':
+                asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
+
+            asyncio.run(self._run_tests(runners))
         finally:
             self.close_logfiles()
 
@@ -2089,10 +2092,6 @@ def run(options: argparse.Namespace) -> int:
 
     if options.wrapper:
         check_bin = options.wrapper[0]
-
-    if sys.platform == 'win32':
-        loop = asyncio.ProactorEventLoop()
-        asyncio.set_event_loop(loop)
 
     if check_bin is not None:
         exe = ExternalProgram(check_bin, silent=True)

--- a/mesonbuild/mtest.py
+++ b/mesonbuild/mtest.py
@@ -502,7 +502,9 @@ class ConsoleLogger(TestLogger):
         self.progress_task = None          # type: T.Optional[asyncio.Future]
         self.max_left_width = 0            # type: int
         self.stop = False
-        self.update = asyncio.Event()
+        # TODO: before 3.10 this cannot be created immediately, because
+        # it will create a new event loop
+        self.update: asyncio.Event
         self.should_erase_line = ''
         self.test_count = 0
         self.started_tests = 0
@@ -600,6 +602,7 @@ class ConsoleLogger(TestLogger):
                 self.emit_progress(harness)
             self.flush()
 
+        self.update = asyncio.Event()
         self.test_count = harness.test_count
         self.cols = max(self.cols, harness.max_left_width + 30)
 

--- a/mesonbuild/mtest.py
+++ b/mesonbuild/mtest.py
@@ -574,7 +574,7 @@ class ConsoleLogger(TestLogger):
 
     def start(self, harness: 'TestHarness') -> None:
         async def report_progress() -> None:
-            loop = asyncio.get_event_loop()
+            loop = asyncio.get_running_loop()
             next_update = 0.0
             self.request_update()
             while not self.stop:
@@ -1232,13 +1232,14 @@ async def complete_all(futures: T.Iterable[asyncio.Future],
 
     # Python is silly and does not have a variant of asyncio.wait with an
     # absolute time as deadline.
-    deadline = None if timeout is None else asyncio.get_event_loop().time() + timeout
+    loop = asyncio.get_running_loop()
+    deadline = None if timeout is None else loop.time() + timeout
     while futures and (timeout is None or timeout > 0):
         done, futures = await asyncio.wait(futures, timeout=timeout,
                                            return_when=asyncio.FIRST_EXCEPTION)
         check_futures(done)
         if deadline:
-            timeout = deadline - asyncio.get_event_loop().time()
+            timeout = deadline - loop.time()
 
     check_futures(futures)
 
@@ -1948,6 +1949,7 @@ class TestHarness:
         running_tests = {}  # type: T.Dict[asyncio.Future, str]
         interrupted = False
         ctrlc_times = deque(maxlen=MAX_CTRLC)  # type: T.Deque[float]
+        loop = asyncio.get_running_loop()
 
         async def run_test(test: SingleTestRunner) -> None:
             async with semaphore:
@@ -1996,7 +1998,7 @@ class TestHarness:
             nonlocal interrupted
             if interrupted:
                 return
-            ctrlc_times.append(asyncio.get_event_loop().time())
+            ctrlc_times.append(loop.time())
             if len(ctrlc_times) == MAX_CTRLC and ctrlc_times[-1] - ctrlc_times[0] < 1:
                 self.flush_logfiles()
                 mlog.warning('CTRL-C detected, exiting')
@@ -2013,10 +2015,10 @@ class TestHarness:
 
         if sys.platform != 'win32':
             if os.getpgid(0) == os.getpid():
-                asyncio.get_event_loop().add_signal_handler(signal.SIGINT, sigint_handler)
+                loop.add_signal_handler(signal.SIGINT, sigint_handler)
             else:
-                asyncio.get_event_loop().add_signal_handler(signal.SIGINT, sigterm_handler)
-            asyncio.get_event_loop().add_signal_handler(signal.SIGTERM, sigterm_handler)
+                loop.add_signal_handler(signal.SIGINT, sigterm_handler)
+            loop.add_signal_handler(signal.SIGTERM, sigterm_handler)
         try:
             for runner in runners:
                 if not runner.is_parallel:
@@ -2033,8 +2035,8 @@ class TestHarness:
             await complete_all(futures)
         finally:
             if sys.platform != 'win32':
-                asyncio.get_event_loop().remove_signal_handler(signal.SIGINT)
-                asyncio.get_event_loop().remove_signal_handler(signal.SIGTERM)
+                loop.remove_signal_handler(signal.SIGINT)
+                loop.remove_signal_handler(signal.SIGTERM)
             for l in self.loggers:
                 await l.finish(self)
 

